### ISSUE TITLE
Rebuilding connections from serializable parts

### ISSUE THREAD

**Issue by jeromegn** [rustls/rustls#1247]
We're looking into passing file descriptors either via `SCM_RIGHTS` or `pidfd_getfd` (probably the latter) techniques when restarting our proxy.

Unfortunately, we have no way to rebuild a TLS connection on the other process, making it impossible to pass encrypted connections.

We tried using `ktls`, but it introduced a few other issues we can't easily overcome.

I'm wondering if it would be possible to:
- Ensure the internal buffer of a connection, if any, is drained (or gain access to the buffer)
- Convert a connection into its parts (something like `Parts<IO>`)
- Serialize the metadata of `Parts<IO>`
- Rebuild a `server::Connection` from `Parts<IO>`

We wouldn't even need a full `server::Connection`, we'd only need something that can still encrypt/decrypt data. We can forward other metadata (like parts of the ClientHello) ourselves if we need to.

This wouldn't be too dissimilar to how `extract_secrets()` work for offloading to `ktls` (I think, but I don't know much).

Is something like this possible within the current design?

**Comment by djc** [comment]
First, note that rustls types don't hold the IO, so there's no `Parts<IO>` here (you'd probably need that from tokio-rustls). Other than that, I think it would probably be feasible to somehow resurrect a limited type of connection from the secrets. I'm assuming we can limit this to connections that are doing traffic (that is, after completion of the handshake). It would take a bit of work, but I'm happy to provide any guidance you need.

**Comment by jeromegn** [comment]
Thanks for clarifying. My bad, I didn't look nearly deep enough.

Right now I still have my hopes up that it won't be necessary to do this work. We're looking into the issues I faced with ktls: https://github.com/hapsoc/ktls/issues/20

It might be useful for others though.

**Comment by djc** [comment]
#1112 added code to extract the secrets from a connection for which the handshake is finished. I imagine we'd be open to a PR that somehow constructs a `ClientConnection` or `ServerConnection` from an `ExtractedSecrets`.

**Comment by ctz** [comment]
> It's really the equivalent of the OpenSSL "d2i_SSL_SESSION" and "i2d_SSL_SESSION" functionality I'm asking about here. Transferring the data/file descriptors between processes is something we'd handle, it just needs a "payload" we can export/import to make the rustls connection state transferrable.

Note that the OpenSSL API you mention also cannot be used to generally transfer an extant connection between processes[^ref]. An OpenSSL `SSL_SESSION` is instead the data to needed to complete a resumed handshake. That happens on a separate underlying TCP connection, so there is no need to think about FD passing either.

If I were you, I'd look into using KTLS (using the API @djc mentioned) and then doing FD passing.

[^ref]: https://mta.openssl.org/pipermail/openssl-users/2021-December/014728.html

**Comment by ctz** [comment]
> I imagine we'd be open to a PR that somehow constructs a `ClientConnection` or `ServerConnection` from an `ExtractedSecrets`.

I don't think this is feasible. An `ExtractedSecrets` would get you a `RecordLayer`. To reconstruct a connection you'd also need (non-exhaustive):

- the TLS1.3 key schedule and exporter secret
- the TLS1.2 master secret and Client/Server randoms
- the peer certificate chain
- the agreed SNI and ALPN (if any)
- the contents of buffers containing read but unprocessed data
- the contents of buffers containing data buffered but unsent messages

**Comment by RogerSanders** [comment]
Thanks for your feedback, it's very much appreciated.

We'll be experimenting with ktls. It is kind of sidestepping the issue, but if it works, that's a win for us. It's Linux specific, which isn't a dealbreaker for our case, but isn't ideal either. You're right that the OpenSSL functions I pointed to aren't enough on their own, and although connection handover between processes can be made to work with OpenSSL, it isn't officially supported. It's probably true that ktls would fall into that same category though.

I'm still interested in the viability of making this work on rustls itself however. This use case is valid, and isn't unique. It seems that if the application (IE, the code above your library) takes responsibility for providing a valid file descriptor, it should be possible for the rustls library to provide a way to "serialize" and "deserialize" a complete active connection. Outside of the OS socket, it's just a bunch of data sitting in buffers. The "serialization" process can/must be "destructive" in this case, it's kind of like "decomposing" the connection into a serialized form, after which that original connection object in the application can no longer be used. A new one must be "recomposed" from a serialized state in order to continue. Apart from the "plumbing" work required to get data out of and back into the various buffers, can you see anything I'm missing here which would prevent this working?
